Extend PICT support

andrews05 opened the pull request:
This PR extends PICT support in a number of ways:
- Adds support for pixmaps with < 8bpp.
- Adds support for v1 PICTs.
- Adds support for bits_rect, bits_region and packbits_region opcodes.
- Adheres to the destination rect to set pixels in the correct location.

[pict-tests.rsrc.zip](https://github.com/TheDiamondProject/Graphite/files/6179340/pict-tests.rsrc.zip)

andrews05 commented:
Do you have any samples of "Old style bitmaps" mentioned in the comments?

tjhancocks commented:
None to hand. The old style bitmaps were from Basic QuickDraw, not Colour QuickDraw. Best source would probably be an ancient MacOS program from around System 1.0 to System 6.0.

tjhancocks commented on libGraphite/quickdraw/pixmap.cpp:
```diff
@@ -155,6 +157,42 @@ auto graphite::qd::pixmap::set_pm_table(const uint32_t& pm_table) -> void
+    const std::vector<uint8_t>& pixel_data,
+    const qd::clut& clut,
+    int64_t offset) -> void
+{
```
Nice. I knew some of these things must be fundamentally the same and sharing code internally, given that certain resources are just specialised instances. 

Great to actually get this happening here 👍

andrews05 commented:
> None to hand. The old style bitmaps were from Basic QuickDraw, not Colour QuickDraw. Best source would probably be an ancient MacOS program from around System 1.0 to System 6.0.

Oh, do you mean v1 PICTs? I have seen some of those, though they're explicitly unsupported at the moment. Does that mean that part of the code is currently unused?

tjhancocks commented:
I think there is some confusion. Probably from me. My understanding is the the bitmap representation of data was used with the basic QuickDraw format, with just black on white coloration (was a while since I wrote this code so apologies for misremembered details). However I may be mistaken here.

andrews05 commented:
Eh, probably both of us 🤷‍♂️
It's fine, I was just wondering how to trigger that code path since I haven't encountered it yet.

andrews05 commented:
Alright, I've done some work on the v1 format (added to this PR as it relies on the existing work done here).
This add the unpacked bits opcode, passed in as a flag to read_indirect_bits (renamed from read_pack_bits).

andrews05 commented:
Conflicts resolved. Does it all look okay?

Combined with the quicktime changes, this can now successfully decode every image in the Galaxy's Edge plugin! I've attached a sample of some of the ones that have had problems.

[ge-pict-tests.rsrc.zip](https://github.com/TheDiamondProject/Graphite/files/6216285/ge-pict-tests.rsrc.zip)

andrews05 commented:
*Sigh* Seems like this never ends. Multiple additional complications with Frozen Heart.
I've added some more out-of-bounds checks to prevent some crashes, added some more opcodes, and throw on unknown opcodes.
And finally figured out how the "Old style bitmaps" come into play :)

[fh-pict-tests.rsrc.zip](https://github.com/TheDiamondProject/Graphite/files/6216743/fh-pict-tests.rsrc.zip)

tjhancocks commented on libGraphite/quickdraw/pict.cpp:
```diff
@@ -448,6 +437,14 @@ auto graphite::qd::pict::parse(graphite::data::reader& pict_reader) -> void
                 break;
             }
+            case opcode::short_comment: {
+                pict_reader.move(2);
```
Might be an idea to copy the style of the `read_long_comment` above, and to actually skip over the contents of the comment if they exist.

andrews05 replied:
As best I could tell from the samples I had and from the imagemagick code, the size is always 2. Does seem like a pretty short comment... 🤔

tjhancocks commented:
Is this essentially ready to merge in your opinion?

andrews05 commented:
Yup, should be good to go.
